{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# DOWNLOAD THE DATASET",
   "id": "cde9ff8b3358a345"
  },
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2026-01-21T13:58:32.392718Z",
     "start_time": "2026-01-21T13:58:30.213450Z"
    }
   },
   "source": [
    "import GEOparse\n",
    "import pandas as pd\n",
    "\n",
    "import os\n",
    "\n",
    "# Create necessary directories\n",
    "os.makedirs('./data', exist_ok=True)\n",
    "os.makedirs('./results', exist_ok=True)\n",
    "os.makedirs('./figures', exist_ok=True)\n",
    "\n",
    "# Download GSE147507 from GEO\n",
    "gse = GEOparse.get_GEO(geo=\"GSE147507\", destdir=\"./data/\")\n",
    "\n",
    "# Examine the dataset\n",
    "print(\"Dataset info:\")\n",
    "print(f\"Title: {gse.metadata['title'][0]}\")\n",
    "print(f\"Samples: {len(gse.gsms)}\")\n",
    "print(\"\\nSample names:\")\n",
    "for gsm_name, gsm in gse.gsms.items():\n",
    "    print(f\"{gsm_name}: {gsm.metadata['title'][0]}\")"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21-Jan-2026 14:58:31 DEBUG utils - Directory ./data/ already exists. Skipping.\n",
      "21-Jan-2026 14:58:31 INFO GEOparse - File already exist: using local version.\n",
      "21-Jan-2026 14:58:31 INFO GEOparse - Parsing ./data/GSE147507_family.soft.gz: \n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SERIES: GSE147507\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - PLATFORM: GPL18573\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - PLATFORM: GPL28369\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432378\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432379\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432380\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432381\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432382\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432383\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432384\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432385\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432386\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432387\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432388\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432389\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432390\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432391\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432392\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432393\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432394\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432395\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432396\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4432397\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462336\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462337\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462338\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462339\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462340\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462341\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462342\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462343\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462344\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462345\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462346\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462347\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462348\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462349\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462350\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462351\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462352\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462353\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462354\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462355\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462356\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462357\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462358\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462359\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462360\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462361\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462362\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462363\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462364\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462365\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462366\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462367\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462368\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462369\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462370\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462371\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462372\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462373\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462374\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462375\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462376\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462377\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462378\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462379\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462380\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462381\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462382\n",
      "21-Jan-2026 14:58:31 DEBUG GEOparse - SAMPLE: GSM4462383\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462384\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462385\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462386\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462387\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462388\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462389\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462390\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462391\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462392\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462393\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462394\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462395\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462396\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462397\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462398\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462399\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462400\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462401\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462402\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462403\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462404\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462405\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462406\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462407\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462408\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462409\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462410\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462411\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462412\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462413\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462414\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462415\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4462416\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4486157\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4486158\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4486159\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4486160\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4486161\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4486162\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4486163\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4486164\n",
      "21-Jan-2026 14:58:32 DEBUG GEOparse - SAMPLE: GSM4486165\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset info:\n",
      "Title: Transcriptional response to SARS-CoV-2 infection\n",
      "Samples: 110\n",
      "\n",
      "Sample names:\n",
      "GSM4432378: Series1_NHBE_Mock_1\n",
      "GSM4432379: Series1_NHBE_Mock_2\n",
      "GSM4432380: Series1_NHBE_Mock_3\n",
      "GSM4432381: Series1_NHBE_SARS-CoV-2_1\n",
      "GSM4432382: Series1_NHBE_SARS-CoV-2_2\n",
      "GSM4432383: Series1_NHBE_SARS-CoV-2_3\n",
      "GSM4432384: Series2_A549_Mock_1\n",
      "GSM4432385: Series2_A549_Mock_2\n",
      "GSM4432386: Series2_A549_Mock_3\n",
      "GSM4432387: Series2_A549_SARS-CoV-2_1\n",
      "GSM4432388: Series2_A549_SARS-CoV-2_2\n",
      "GSM4432389: Series2_A549_SARS-CoV-2_3\n",
      "GSM4432390: Series3_A549_Mock_1\n",
      "GSM4432391: Series3_A549_Mock_2\n",
      "GSM4432392: Series3_A549_RSV_1\n",
      "GSM4432393: Series3_A549_RSV_2\n",
      "GSM4432394: Series4_A549_Mock_1\n",
      "GSM4432395: Series4_A549_Mock_2\n",
      "GSM4432396: Series4_A549_IAV_1\n",
      "GSM4432397: Series4_A549_IAV_2\n",
      "GSM4462336: Series5_A549_Mock_1\n",
      "GSM4462337: Series5_A549_Mock_2\n",
      "GSM4462338: Series5_A549_Mock_3\n",
      "GSM4462339: Series5_A549_SARS-CoV-2_1\n",
      "GSM4462340: Series5_A549_SARS-CoV-2_2\n",
      "GSM4462341: Series5_A549_SARS-CoV-2_3\n",
      "GSM4462342: Series6_A549-ACE2_Mock_1\n",
      "GSM4462343: Series6_A549-ACE2_Mock_2\n",
      "GSM4462344: Series6_A549-ACE2_Mock_3\n",
      "GSM4462345: Series6_A549-ACE2_SARS-CoV-2_1\n",
      "GSM4462346: Series6_A549-ACE2_SARS-CoV-2_2\n",
      "GSM4462347: Series6_A549-ACE2_SARS-CoV-2_3\n",
      "GSM4462348: Series7_Calu3_Mock_1\n",
      "GSM4462349: Series7_Calu3_Mock_2\n",
      "GSM4462350: Series7_Calu3_Mock_3\n",
      "GSM4462351: Series7_Calu3_SARS-CoV-2_1\n",
      "GSM4462352: Series7_Calu3_SARS-CoV-2_2\n",
      "GSM4462353: Series7_Calu3_SARS-CoV-2_3\n",
      "GSM4462354: Series8_A549_Mock_1\n",
      "GSM4462355: Series8_A549_Mock_2\n",
      "GSM4462356: Series8_A549_Mock_3\n",
      "GSM4462357: Series8_A549_RSV_1\n",
      "GSM4462358: Series8_A549_RSV_2\n",
      "GSM4462359: Series8_A549_RSV_3\n",
      "GSM4462360: Series8_A549_HPIV3_1\n",
      "GSM4462361: Series8_A549_HPIV3_2\n",
      "GSM4462362: Series8_A549_HPIV3_3\n",
      "GSM4462363: Series9_NHBE_Mock_1\n",
      "GSM4462364: Series9_NHBE_Mock_2\n",
      "GSM4462365: Series9_NHBE_Mock_3\n",
      "GSM4462366: Series9_NHBE_Mock_4\n",
      "GSM4462367: Series9_NHBE_IAV_1\n",
      "GSM4462368: Series9_NHBE_IAV_2\n",
      "GSM4462369: Series9_NHBE_IAV_3\n",
      "GSM4462370: Series9_NHBE_IAV_4\n",
      "GSM4462371: Series9_NHBE_IAVdNS1_1\n",
      "GSM4462372: Series9_NHBE_IAVdNS1_2\n",
      "GSM4462373: Series9_NHBE_IAVdNS1_3\n",
      "GSM4462374: Series9_NHBE_IAVdNS1_4\n",
      "GSM4462375: Series9_NHBE_IFNB_4h_1\n",
      "GSM4462376: Series9_NHBE_IFNB_4h_2\n",
      "GSM4462377: Series9_NHBE_IFNB_6h_1\n",
      "GSM4462378: Series9_NHBE_IFNB_6h_2\n",
      "GSM4462379: Series9_NHBE_IFNB_12h_1\n",
      "GSM4462380: Series9_NHBE_IFNB_12h_2\n",
      "GSM4462381: Series10_FerretNW_Ctl_d1_1\n",
      "GSM4462382: Series10_FerretNW_Ctl_d1_2\n",
      "GSM4462383: Series10_FerretNW_SARS-CoV-2_d1_1\n",
      "GSM4462384: Series10_FerretNW_SARS-CoV-2_d1_2\n",
      "GSM4462385: Series11_FerretNW_Ctl_d3_1\n",
      "GSM4462386: Series11_FerretNW_Ctl_d3_2\n",
      "GSM4462387: Series11_FerretNW_SARS-CoV-2_d3_1\n",
      "GSM4462388: Series11_FerretNW_SARS-CoV-2_d3_2\n",
      "GSM4462389: Series12_FerretNW_Ctl_d7_1\n",
      "GSM4462390: Series12_FerretNW_Ctl_d7_2\n",
      "GSM4462391: Series12_FerretNW_SARS-CoV-2_d7_1\n",
      "GSM4462392: Series12_FerretNW_SARS-CoV-2_d7_2\n",
      "GSM4462393: Series12_FerretNW_IAV_d7_1\n",
      "GSM4462394: Series12_FerretNW_IAV_d7_2\n",
      "GSM4462395: Series13_FerretNW_Ctl_d14_1\n",
      "GSM4462396: Series13_FerretNW_Ctl_d14_2\n",
      "GSM4462397: Series13_FerretNW_SARS-CoV-2_d14_1\n",
      "GSM4462398: Series13_FerretNW_SARS-CoV-2_d14_2\n",
      "GSM4462399: Series14_FerretTrachea_Ctl_d3_1\n",
      "GSM4462400: Series14_FerretTrachea_Ctl_d3_2\n",
      "GSM4462401: Series14_FerretTrachea_Ctl_d3_3\n",
      "GSM4462402: Series14_FerretTrachea_Ctl_d3_4\n",
      "GSM4462403: Series14_FerretTrachea_SARS-CoV-2_d3_1\n",
      "GSM4462404: Series14_FerretTrachea_SARS-CoV-2_d3_2\n",
      "GSM4462405: Series14_FerretTrachea_SARS-CoV-2_d3_3\n",
      "GSM4462406: Series14_FerretTrachea_SARS-CoV-2_d3_4\n",
      "GSM4462407: Series14_FerretTrachea_IAV_d3_1\n",
      "GSM4462408: Series14_FerretTrachea_IAV_d3_2\n",
      "GSM4462409: Series14_FerretTrachea_IAV_d3_3\n",
      "GSM4462410: Series14_FerretTrachea_IAV_d3_4\n",
      "GSM4462411: Series14_FerretTrachea_IAV_d3_5\n",
      "GSM4462412: Series14_FerretTrachea_IAV_d3_6\n",
      "GSM4462413: Series15_HealthyLungBiopsy_1\n",
      "GSM4462414: Series15_HealthyLungBiopsy_2\n",
      "GSM4462415: Series15_COVID19Lung_1\n",
      "GSM4462416: Series15_COVID19Lung_2\n",
      "GSM4486157: Series16_A549-ACE2_Mock_1\n",
      "GSM4486158: Series16_A549-ACE2_Mock_2\n",
      "GSM4486159: Series16_A549-ACE2_Mock_3\n",
      "GSM4486160: Series16_A549-ACE2_SARS-CoV-2_1\n",
      "GSM4486161: Series16_A549-ACE2_SARS-CoV-2_2\n",
      "GSM4486162: Series16_A549-ACE2_SARS-CoV-2_3\n",
      "GSM4486163: Series16_A549-ACE2_SARS-CoV-2_Rux_1\n",
      "GSM4486164: Series16_A549-ACE2_SARS-CoV-2_Rux_2\n",
      "GSM4486165: Series16_A549-ACE2_SARS-CoV-2_Rux_3\n"
     ]
    }
   ],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-21T13:58:33.007613Z",
     "start_time": "2026-01-21T13:58:32.991106Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Check how data is stored in this dataset\n",
    "first_sample = list(gse.gsms.values())[0]\n",
    "print(\"Sample metadata keys:\", first_sample.metadata.keys())\n",
    "print(\"\\nChecking for supplementary files:\")\n",
    "print(\"Supplementary files:\", first_sample.metadata.get('supplementary_file', 'None'))\n",
    "print(\"\\nData table type:\", type(first_sample.table))\n",
    "print(\"Columns:\", first_sample.table.columns.tolist() if hasattr(first_sample.table, 'columns') else 'No columns')\n",
    "\n",
    "# Check the GSE level\n",
    "print(\"\\nGSE supplementary files:\")\n",
    "print(gse.metadata.get('supplementary_file', 'None'))"
   ],
   "id": "d6e59581803224f8",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample metadata keys: dict_keys(['title', 'geo_accession', 'status', 'submission_date', 'last_update_date', 'type', 'channel_count', 'source_name_ch1', 'organism_ch1', 'taxid_ch1', 'characteristics_ch1', 'molecule_ch1', 'extract_protocol_ch1', 'description', 'data_processing', 'platform_id', 'contact_name', 'contact_laboratory', 'contact_department', 'contact_institute', 'contact_address', 'contact_city', 'contact_state', 'contact_zip/postal_code', 'contact_country', 'instrument_model', 'library_selection', 'library_source', 'library_strategy', 'relation', 'supplementary_file_1', 'series_id', 'data_row_count'])\n",
      "\n",
      "Checking for supplementary files:\n",
      "Supplementary files: None\n",
      "\n",
      "Data table type: <class 'pandas.core.frame.DataFrame'>\n",
      "Columns: []\n",
      "\n",
      "GSE supplementary files:\n",
      "['ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE147nnn/GSE147507/suppl/GSE147507_RawReadCounts_Ferret.tsv.gz', 'ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE147nnn/GSE147507/suppl/GSE147507_RawReadCounts_Human.tsv.gz']\n"
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Load and Prepare the Data",
   "id": "4321fe4c9947b86e"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-21T14:02:13.783158Z",
     "start_time": "2026-01-21T14:02:13.394889Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import urllib.request\n",
    "import gzip\n",
    "import shutil\n",
    "\n",
    "# Download the raw counts file\n",
    "print(\"Downloading raw count data...\")\n",
    "url = \"ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE147nnn/GSE147507/suppl/GSE147507_RawReadCounts_Human.tsv.gz\"\n",
    "local_gz = \"./data/raw_counts.tsv.gz\"\n",
    "local_tsv = \"./data/raw_counts.tsv\"\n",
    "\n",
    "# Download\n",
    "urllib.request.urlretrieve(url, local_gz)\n",
    "\n",
    "# Unzip\n",
    "with gzip.open(local_gz, 'rb') as f_in:\n",
    "    with open(local_tsv, 'wb') as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)\n",
    "\n",
    "print(\"Loading expression data...\")\n",
    "expr_df = pd.read_csv(local_tsv, sep='\\t', index_col=0)\n",
    "\n",
    "print(f\"Expression matrix shape: {expr_df.shape}\")\n",
    "print(f\"\\nColumn names (samples):\\n{expr_df.columns.tolist()}\")\n",
    "print(f\"\\nFirst few rows:\\n{expr_df.head()}\")\n",
    "\n",
    "# Create sample info based on column names\n",
    "sample_info = []\n",
    "for col in expr_df.columns:\n",
    "    if 'SARS' in col or 'CoV' in col:\n",
    "        condition = 'Infected'\n",
    "    elif 'Mock' in col or 'Series' in col:\n",
    "        condition = 'Mock'\n",
    "    else:\n",
    "        condition = 'Unknown'\n",
    "    sample_info.append({'Sample': col, 'Condition': condition})\n",
    "\n",
    "sample_df = pd.DataFrame(sample_info)\n",
    "print(f\"\\nSample conditions:\\n{sample_df}\")\n",
    "\n",
    "# Save\n",
    "expr_df.to_csv('./data/raw_expression.csv')\n",
    "sample_df.to_csv('./data/sample_info.csv', index=False)"
   ],
   "id": "8b574a43510be78b",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading raw count data...\n"
     ]
    },
    {
     "ename": "URLError",
     "evalue": "<urlopen error [Errno 11002] getaddrinfo failed>",
     "output_type": "error",
     "traceback": [
      "\u001B[31m---------------------------------------------------------------------------\u001B[39m",
      "\u001B[31mgaierror\u001B[39m                                  Traceback (most recent call last)",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\urllib\\request.py:1531\u001B[39m, in \u001B[36mFTPHandler.ftp_open\u001B[39m\u001B[34m(self, req)\u001B[39m\n\u001B[32m   1530\u001B[39m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[32m-> \u001B[39m\u001B[32m1531\u001B[39m     host = \u001B[43msocket\u001B[49m\u001B[43m.\u001B[49m\u001B[43mgethostbyname\u001B[49m\u001B[43m(\u001B[49m\u001B[43mhost\u001B[49m\u001B[43m)\u001B[49m\n\u001B[32m   1532\u001B[39m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mOSError\u001B[39;00m \u001B[38;5;28;01mas\u001B[39;00m msg:\n",
      "\u001B[31mgaierror\u001B[39m: [Errno 11002] getaddrinfo failed",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001B[31mURLError\u001B[39m                                  Traceback (most recent call last)",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[4]\u001B[39m\u001B[32m, line 18\u001B[39m\n\u001B[32m     15\u001B[39m local_tsv = \u001B[33m\"\u001B[39m\u001B[33m./data/raw_counts.tsv\u001B[39m\u001B[33m\"\u001B[39m\n\u001B[32m     17\u001B[39m \u001B[38;5;66;03m# Download\u001B[39;00m\n\u001B[32m---> \u001B[39m\u001B[32m18\u001B[39m \u001B[43murllib\u001B[49m\u001B[43m.\u001B[49m\u001B[43mrequest\u001B[49m\u001B[43m.\u001B[49m\u001B[43murlretrieve\u001B[49m\u001B[43m(\u001B[49m\u001B[43murl\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mlocal_gz\u001B[49m\u001B[43m)\u001B[49m\n\u001B[32m     20\u001B[39m \u001B[38;5;66;03m# Unzip\u001B[39;00m\n\u001B[32m     21\u001B[39m \u001B[38;5;28;01mwith\u001B[39;00m gzip.open(local_gz, \u001B[33m'\u001B[39m\u001B[33mrb\u001B[39m\u001B[33m'\u001B[39m) \u001B[38;5;28;01mas\u001B[39;00m f_in:\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\urllib\\request.py:214\u001B[39m, in \u001B[36murlretrieve\u001B[39m\u001B[34m(url, filename, reporthook, data)\u001B[39m\n\u001B[32m    197\u001B[39m \u001B[38;5;250m\u001B[39m\u001B[33;03m\"\"\"\u001B[39;00m\n\u001B[32m    198\u001B[39m \u001B[33;03mRetrieve a URL into a temporary location on disk.\u001B[39;00m\n\u001B[32m    199\u001B[39m \n\u001B[32m   (...)\u001B[39m\u001B[32m    210\u001B[39m \u001B[33;03mdata file as well as the resulting HTTPMessage object.\u001B[39;00m\n\u001B[32m    211\u001B[39m \u001B[33;03m\"\"\"\u001B[39;00m\n\u001B[32m    212\u001B[39m url_type, path = _splittype(url)\n\u001B[32m--> \u001B[39m\u001B[32m214\u001B[39m \u001B[38;5;28;01mwith\u001B[39;00m contextlib.closing(\u001B[43murlopen\u001B[49m\u001B[43m(\u001B[49m\u001B[43murl\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdata\u001B[49m\u001B[43m)\u001B[49m) \u001B[38;5;28;01mas\u001B[39;00m fp:\n\u001B[32m    215\u001B[39m     headers = fp.info()\n\u001B[32m    217\u001B[39m     \u001B[38;5;66;03m# Just return the local path and the \"headers\" for file://\u001B[39;00m\n\u001B[32m    218\u001B[39m     \u001B[38;5;66;03m# URLs. No sense in performing a copy unless requested.\u001B[39;00m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\urllib\\request.py:189\u001B[39m, in \u001B[36murlopen\u001B[39m\u001B[34m(url, data, timeout, context)\u001B[39m\n\u001B[32m    187\u001B[39m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[32m    188\u001B[39m     opener = _opener\n\u001B[32m--> \u001B[39m\u001B[32m189\u001B[39m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mopener\u001B[49m\u001B[43m.\u001B[49m\u001B[43mopen\u001B[49m\u001B[43m(\u001B[49m\u001B[43murl\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdata\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mtimeout\u001B[49m\u001B[43m)\u001B[49m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\urllib\\request.py:489\u001B[39m, in \u001B[36mOpenerDirector.open\u001B[39m\u001B[34m(self, fullurl, data, timeout)\u001B[39m\n\u001B[32m    486\u001B[39m     req = meth(req)\n\u001B[32m    488\u001B[39m sys.audit(\u001B[33m'\u001B[39m\u001B[33murllib.Request\u001B[39m\u001B[33m'\u001B[39m, req.full_url, req.data, req.headers, req.get_method())\n\u001B[32m--> \u001B[39m\u001B[32m489\u001B[39m response = \u001B[38;5;28;43mself\u001B[39;49m\u001B[43m.\u001B[49m\u001B[43m_open\u001B[49m\u001B[43m(\u001B[49m\u001B[43mreq\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdata\u001B[49m\u001B[43m)\u001B[49m\n\u001B[32m    491\u001B[39m \u001B[38;5;66;03m# post-process response\u001B[39;00m\n\u001B[32m    492\u001B[39m meth_name = protocol+\u001B[33m\"\u001B[39m\u001B[33m_response\u001B[39m\u001B[33m\"\u001B[39m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\urllib\\request.py:506\u001B[39m, in \u001B[36mOpenerDirector._open\u001B[39m\u001B[34m(self, req, data)\u001B[39m\n\u001B[32m    503\u001B[39m     \u001B[38;5;28;01mreturn\u001B[39;00m result\n\u001B[32m    505\u001B[39m protocol = req.type\n\u001B[32m--> \u001B[39m\u001B[32m506\u001B[39m result = \u001B[38;5;28;43mself\u001B[39;49m\u001B[43m.\u001B[49m\u001B[43m_call_chain\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[43m.\u001B[49m\u001B[43mhandle_open\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mprotocol\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mprotocol\u001B[49m\u001B[43m \u001B[49m\u001B[43m+\u001B[49m\n\u001B[32m    507\u001B[39m \u001B[43m                          \u001B[49m\u001B[33;43m'\u001B[39;49m\u001B[33;43m_open\u001B[39;49m\u001B[33;43m'\u001B[39;49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mreq\u001B[49m\u001B[43m)\u001B[49m\n\u001B[32m    508\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m result:\n\u001B[32m    509\u001B[39m     \u001B[38;5;28;01mreturn\u001B[39;00m result\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\urllib\\request.py:466\u001B[39m, in \u001B[36mOpenerDirector._call_chain\u001B[39m\u001B[34m(self, chain, kind, meth_name, *args)\u001B[39m\n\u001B[32m    464\u001B[39m \u001B[38;5;28;01mfor\u001B[39;00m handler \u001B[38;5;129;01min\u001B[39;00m handlers:\n\u001B[32m    465\u001B[39m     func = \u001B[38;5;28mgetattr\u001B[39m(handler, meth_name)\n\u001B[32m--> \u001B[39m\u001B[32m466\u001B[39m     result = \u001B[43mfunc\u001B[49m\u001B[43m(\u001B[49m\u001B[43m*\u001B[49m\u001B[43margs\u001B[49m\u001B[43m)\u001B[49m\n\u001B[32m    467\u001B[39m     \u001B[38;5;28;01mif\u001B[39;00m result \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[32m    468\u001B[39m         \u001B[38;5;28;01mreturn\u001B[39;00m result\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\urllib\\request.py:1533\u001B[39m, in \u001B[36mFTPHandler.ftp_open\u001B[39m\u001B[34m(self, req)\u001B[39m\n\u001B[32m   1531\u001B[39m     host = socket.gethostbyname(host)\n\u001B[32m   1532\u001B[39m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mOSError\u001B[39;00m \u001B[38;5;28;01mas\u001B[39;00m msg:\n\u001B[32m-> \u001B[39m\u001B[32m1533\u001B[39m     \u001B[38;5;28;01mraise\u001B[39;00m URLError(msg)\n\u001B[32m   1534\u001B[39m path, attrs = _splitattr(req.selector)\n\u001B[32m   1535\u001B[39m dirs = path.split(\u001B[33m'\u001B[39m\u001B[33m/\u001B[39m\u001B[33m'\u001B[39m)\n",
      "\u001B[31mURLError\u001B[39m: <urlopen error [Errno 11002] getaddrinfo failed>"
     ]
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Quality Control & Normalization",
   "id": "e2f80d448d3ee02e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Filter low-expressed genes (must have >10 reads in at least 3 samples)\n",
    "expr_filtered = expr_df[(expr_df > 10).sum(axis=1) >= 3]\n",
    "print(f\"Genes after filtering: {expr_filtered.shape[0]}\")\n",
    "\n",
    "# Log2 transform (add pseudocount to avoid log(0))\n",
    "expr_log2 = np.log2(expr_filtered + 1)\n",
    "\n",
    "# Create QC plots\n",
    "fig, axes = plt.subplots(1, 2, figsize=(12, 5))\n",
    "\n",
    "# Boxplot of distributions\n",
    "expr_log2.boxplot(ax=axes[0], rot=90)\n",
    "axes[0].set_title('Expression Distribution (log2)')\n",
    "axes[0].set_ylabel('log2(counts + 1)')\n",
    "\n",
    "# Sample correlation heatmap\n",
    "corr_matrix = expr_log2.corr()\n",
    "sns.heatmap(corr_matrix, annot=True, fmt='.2f', cmap='coolwarm', ax=axes[1])\n",
    "axes[1].set_title('Sample Correlation')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('./figures/qc_plots.png', dpi=300, bbox_inches='tight')\n",
    "print(\"QC plots saved!\")"
   ],
   "id": "a328efa7085cbd9c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Differential Expression Analysis",
   "id": "d3490db540b3d1d0"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# CORRECTED: Separate infected vs mock samples\n",
    "# Only \"Mock\" samples go to mock, only \"SARS-CoV-2\" (without Mock) go to infected\n",
    "mock_cols = [col for col in expr_filtered.columns if 'Mock' in col]\n",
    "infected_cols = [col for col in expr_filtered.columns if 'SARS-CoV-2' in col and col not in mock_cols]\n",
    "\n",
    "print(f\"\\nInfected samples ({len(infected_cols)}): {infected_cols}\")\n",
    "print(f\"\\nMock samples ({len(mock_cols)}): {mock_cols}\")\n",
    "\n",
    "# Verify no overlap\n",
    "overlap = set(infected_cols) & set(mock_cols)\n",
    "assert len(overlap) == 0, f\"ERROR: Samples in both groups: {overlap}\"\n",
    "print(\"âœ“ Samples correctly separated!\\n\")\n",
    "\n",
    "infected = expr_filtered[infected_cols]\n",
    "mock = expr_filtered[mock_cols]\n",
    "\n",
    "# Calculate statistics for each gene\n",
    "de_results = []\n",
    "\n",
    "for gene in expr_filtered.index:\n",
    "    infected_vals = infected.loc[gene].values\n",
    "    mock_vals = mock.loc[gene].values\n",
    "\n",
    "    # Calculate means with pseudocount\n",
    "    mean_infected = np.mean(infected_vals) + 1\n",
    "    mean_mock = np.mean(mock_vals) + 1\n",
    "\n",
    "    fold_change = mean_infected / mean_mock\n",
    "    log2fc = np.log2(fold_change)\n",
    "\n",
    "    # T-test\n",
    "    t_stat, p_value = stats.ttest_ind(infected_vals, mock_vals)\n",
    "\n",
    "    de_results.append({\n",
    "        'Gene': gene,\n",
    "        'Mean_Infected': mean_infected,\n",
    "        'Mean_Mock': mean_mock,\n",
    "        'FoldChange': fold_change,\n",
    "        'Log2FC': log2fc,\n",
    "        'P_value': p_value\n",
    "    })\n",
    "\n",
    "# Create results dataframe\n",
    "de_df = pd.DataFrame(de_results)\n",
    "\n",
    "# Remove any infinite or NaN values\n",
    "de_df = de_df.replace([np.inf, -np.inf], np.nan).dropna()\n",
    "print(f\"Genes after removing inf/nan: {len(de_df)}\")\n",
    "\n",
    "# Adjust p-values\n",
    "de_df['Adjusted_P'] = multipletests(de_df['P_value'], method='fdr_bh')[1]\n",
    "de_df['-log10(p)'] = -np.log10(de_df['Adjusted_P'])\n",
    "\n",
    "# Classify genes\n",
    "de_df['Regulation'] = 'Not Significant'\n",
    "de_df.loc[(de_df['Log2FC'] > 1) & (de_df['Adjusted_P'] < 0.05), 'Regulation'] = 'Up'\n",
    "de_df.loc[(de_df['Log2FC'] < -1) & (de_df['Adjusted_P'] < 0.05), 'Regulation'] = 'Down'\n",
    "\n",
    "print(\"\\n\" + \"=\"*50)\n",
    "print(\"DIFFERENTIAL EXPRESSION RESULTS\")\n",
    "print(\"=\"*50)\n",
    "print(de_df['Regulation'].value_counts())\n",
    "print(f\"\\nTop 10 upregulated:\")\n",
    "print(de_df[de_df['Regulation'] == 'Up'].nlargest(10, 'Log2FC')[['Gene', 'Log2FC', 'Adjusted_P']])\n",
    "\n",
    "de_df.to_csv('./results/differential_expression_results.csv', index=False)\n",
    "print(\"\\nâœ… Results saved!\")"
   ],
   "id": "eb0548d586583462",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Check your DE results\n",
    "print(\"\\n\" + \"=\"*60)\n",
    "print(\"YOUR FINAL DIFFERENTIAL EXPRESSION RESULTS\")\n",
    "print(\"=\"*60)\n",
    "print(f\"Total genes analyzed: {len(de_df)}\")\n",
    "print(f\"\\nDifferentially Expressed Genes (|Log2FC| > 1, adj.p < 0.05):\")\n",
    "print(de_df['Regulation'].value_counts())\n",
    "print(f\"\\nðŸ“Š SUMMARY:\")\n",
    "print(f\"   Upregulated: {sum(de_df['Regulation'] == 'Up')} genes\")\n",
    "print(f\"   Downregulated: {sum(de_df['Regulation'] == 'Down')} genes\")\n",
    "print(f\"   Not significant: {sum(de_df['Regulation'] == 'Not Significant')} genes\")"
   ],
   "id": "c75bd32095869588",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Check genes with negative fold changes\n",
    "print(f\"\\nGenes with Log2FC < -1: {sum(de_df['Log2FC'] < -1)}\")\n",
    "print(f\"Genes with Log2FC < -1 AND adj.p < 0.05: {sum((de_df['Log2FC'] < -1) & (de_df['Adjusted_P'] < 0.05))}\")\n",
    "print(f\"\\nMost downregulated genes (even if not significant):\")\n",
    "print(de_df.nsmallest(10, 'Log2FC')[['Gene', 'Log2FC', 'Adjusted_P', 'Regulation']])"
   ],
   "id": "aebdfba9efb3f0c0",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Debug checks\n",
    "print(f\"\\nDE Results shape: {de_df.shape}\")\n",
    "print(f\"Log2FC range: {de_df['Log2FC'].min()} to {de_df['Log2FC'].max()}\")\n",
    "print(f\"P-value range: {de_df['P_value'].min()} to {de_df['P_value'].max()}\")\n",
    "print(f\"-log10(p) range: {de_df['-log10(p)'].min()} to {de_df['-log10(p)'].max()}\")\n",
    "print(f\"\\nAny inf or nan values?\")\n",
    "print(f\"Log2FC inf: {np.isinf(de_df['Log2FC']).sum()}\")\n",
    "print(f\"Log2FC nan: {de_df['Log2FC'].isna().sum()}\")\n",
    "print(f\"-log10(p) inf: {np.isinf(de_df['-log10(p)']).sum()}\")\n",
    "print(f\"-log10(p) nan: {de_df['-log10(p)'].isna().sum()}\")\n",
    "print(f\"\\nFirst few rows of DE results:\")\n",
    "print(de_df[['Gene', 'Log2FC', 'P_value', 'Adjusted_P', '-log10(p)', 'Regulation']].head(10))"
   ],
   "id": "db3e0b1160e727d3",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Volcano Plot",
   "id": "3b7f83ca72f99e9b"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-21T14:09:35.172351Z",
     "start_time": "2026-01-21T14:09:33.654Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from adjustText import adjust_text\n",
    "\n",
    "# Create volcano plot with better scaling\n",
    "fig, ax = plt.subplots(figsize=(12, 10))\n",
    "\n",
    "# Define colors\n",
    "colors = {'Up': '#FF6B6B', 'Down': '#4ECDC4', 'Not Significant': '#95A5A6'}\n",
    "\n",
    "# Plot each group\n",
    "for regulation in ['Not Significant', 'Down', 'Up']:  # Plot non-sig first so sig points are on top\n",
    "    subset = de_df[de_df['Regulation'] == regulation]\n",
    "    ax.scatter(subset['Log2FC'], subset['-log10(p)'],\n",
    "               c=colors[regulation], label=regulation, alpha=0.6, s=30,\n",
    "               edgecolors='none')\n",
    "\n",
    "# Add threshold lines\n",
    "ax.axhline(y=-np.log10(0.05), color='black', linestyle='--', linewidth=1.5, label='p=0.05')\n",
    "ax.axvline(x=1, color='black', linestyle='--', linewidth=1.5, alpha=0.7)\n",
    "ax.axvline(x=-1, color='black', linestyle='--', linewidth=1.5, alpha=0.7)\n",
    "\n",
    "# Set reasonable axis limits\n",
    "ax.set_xlim(de_df['Log2FC'].quantile(0.01) - 0.5, de_df['Log2FC'].quantile(0.99) + 0.5)\n",
    "ax.set_ylim(-0.5, de_df['-log10(p)'].quantile(0.99) + 2)\n",
    "\n",
    "# Labels and styling\n",
    "ax.set_xlabel('logâ‚‚(Fold Change)', fontsize=14, fontweight='bold')\n",
    "ax.set_ylabel('-logâ‚â‚€(Adjusted P-value)', fontsize=14, fontweight='bold')\n",
    "ax.set_title('SARS-CoV-2 Differential Gene Expression\\nInfected vs Mock',\n",
    "             fontsize=16, fontweight='bold', pad=20)\n",
    "ax.legend(loc='upper right', frameon=True, fontsize=11)\n",
    "ax.grid(True, alpha=0.2, linestyle=':')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('./figures/volcano_plot.png', dpi=300, bbox_inches='tight')\n",
    "print(f\"\\nâœ… Volcano plot saved! Found {sum(de_df['Regulation']=='Up')} upregulated and {sum(de_df['Regulation']=='Down')} downregulated genes!\")\n",
    "plt.show()"
   ],
   "id": "f1fc057e48d624ee",
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'de_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001B[31m---------------------------------------------------------------------------\u001B[39m",
      "\u001B[31mNameError\u001B[39m                                 Traceback (most recent call last)",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[5]\u001B[39m\u001B[32m, line 11\u001B[39m\n\u001B[32m      9\u001B[39m \u001B[38;5;66;03m# Plot each group\u001B[39;00m\n\u001B[32m     10\u001B[39m \u001B[38;5;28;01mfor\u001B[39;00m regulation \u001B[38;5;129;01min\u001B[39;00m [\u001B[33m'\u001B[39m\u001B[33mNot Significant\u001B[39m\u001B[33m'\u001B[39m, \u001B[33m'\u001B[39m\u001B[33mDown\u001B[39m\u001B[33m'\u001B[39m, \u001B[33m'\u001B[39m\u001B[33mUp\u001B[39m\u001B[33m'\u001B[39m]:  \u001B[38;5;66;03m# Plot non-sig first so sig points are on top\u001B[39;00m\n\u001B[32m---> \u001B[39m\u001B[32m11\u001B[39m     subset = \u001B[43mde_df\u001B[49m[de_df[\u001B[33m'\u001B[39m\u001B[33mRegulation\u001B[39m\u001B[33m'\u001B[39m] == regulation]\n\u001B[32m     12\u001B[39m     ax.scatter(subset[\u001B[33m'\u001B[39m\u001B[33mLog2FC\u001B[39m\u001B[33m'\u001B[39m], subset[\u001B[33m'\u001B[39m\u001B[33m-log10(p)\u001B[39m\u001B[33m'\u001B[39m],\n\u001B[32m     13\u001B[39m                c=colors[regulation], label=regulation, alpha=\u001B[32m0.6\u001B[39m, s=\u001B[32m30\u001B[39m,\n\u001B[32m     14\u001B[39m                edgecolors=\u001B[33m'\u001B[39m\u001B[33mnone\u001B[39m\u001B[33m'\u001B[39m)\n\u001B[32m     16\u001B[39m \u001B[38;5;66;03m# Add threshold lines\u001B[39;00m\n",
      "\u001B[31mNameError\u001B[39m: name 'de_df' is not defined"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 1200x1000 with 1 Axes>"
      ],
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA9wAAAMzCAYAAABUQhpKAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjgsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvwVt1zgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAKLdJREFUeJzt3XtslYX5wPGnlXBRQGkrZNOoGXhBREVYXLImS+ZwaHQCToO6oRPEZRFNdNOIUfA2Fc0SlSVeli4YyaLE23RIFJ3/eM1QIGgweImX6BzXSOQ20vPL+27tz8p0nNGHlnM+n6S25+176As+LXzPe2uoVCqVAAAAALpVY/f+cgAAAIDgBgAAgCT2cAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAA9Kbg3r59e5x22mnx6quvfu06b731Vpx11llx3HHHxZlnnhkrV678X78cAAAA1H5wb9u2LS6//PJYvXr1166zefPmmDFjRowbNy4effTRGDNmTFx88cXlcgAAAKh1VQf3O++8E2effXZ8+OGH37jeokWLol+/fnHllVfG8OHD45prron99tsvFi9evDvbCwAAALUZ3K+99lqceOKJ8dBDD33jesuXL4+xY8dGQ0ND+bh4f8IJJ8SyZcv+960FAACAvUSfap9w7rnn7tJ6a9asiREjRnRZ1tzc/I2HoQMAAECtSLtK+ZYtW6Jv375dlhWPi4utAQAAQK2reg/3rirO3/5qXBeP+/fvX9Wvs379pqhUunnjoJcozrhoahpkzqlp5px6YM6pB+aceprzXh/cw4YNi7Vr13ZZVjweOnRoVb9OEdvt7d28cdBL/PsSB+WMe2GJWmXOqQfmnHpgzqkHjY17ySHlxb2333jjjaj8uyKK96+//nq5HAAAAGpdtwZ3caG0rVu3lh9PmDAhPv/887j55pvLW4kV74vzuk855ZTu/JIAAABQ+8Hd2tpa3n+7MHDgwLj33ntj6dKlMXny5PI2Yffdd1/su+++3fklAQAAoFdqqHQc891LrVu3yTnc1PS5UC0tg2LtWhcHpHaZc+qBOacemHPq5Rzu5uZBvf8cbgAAAKhnghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAACCB4AYAAIAEghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAACCB4AYAAIAEghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAAEBwAwAAwN7BHm4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAACCB4AYAAIAEghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAACCB4AYAAIAEghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAACCB4AYAAIAEghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAACCB4AYAAIAEghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAACCB4AYAAIAEghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAACCB4AYAAIAEghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAACCB4AYAAIAEghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAACCB4AYAAIAEghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAOgNwb1t27aYNWtWjBs3LlpbW6Otre1r13322WfjlFNOiTFjxsQ555wTb7755u5uLwAAANRmcM+dOzdWrlwZ8+fPj9mzZ8e8efNi8eLFO623evXquOKKK+Liiy+OJ554IkaOHFl+vGXLlu7adgAAAKiN4N68eXMsXLgwrrnmmhg1alSMHz8+pk+fHgsWLNhp3RdffDFGjBgREydOjEMOOSQuv/zyWLNmTbzzzjvduf0AAACw9wf3qlWrYseOHeUh4h3Gjh0by5cvj/b29i7rHnDAAWVcL126tPzco48+GgMHDizjGwAAAGpdn2pWLvZQDxkyJPr27du5rKWlpTyve+PGjdHU1NS5/NRTT43nn38+zj333Nhnn32isbEx7r333th///2r2sCGhn+9QS3qmG0zTi0z59QDc049MOfUg4aGHgzu4vzrL8d2oePx9u3buyzfsGFDGejXXXddHHfccfGnP/0prr766njssceiubl5l79mU9OgajYR9krNzeac2mfOqQfmnHpgziEpuPv167dTWHc87t+/f5fld9xxRxxxxBFx3nnnlY9vvPHG8orljzzySMyYMWOXv+b69ZviK0erQ029glb8pbVu3aaoVHp6ayCHOacemHPqgTmnHjQ2du9O36qCe9iwYeWe6+I87j59/vXUYi92EduDBw/usm5xC7Cf//znnY+LQ8qPOuqo+OSTT6rawCJChAi1zpxTD8w59cCcUw/MObWsUunBi6YVt/YqQnvZsmWdy4qLoo0ePboM6i8bOnRovPvuu12Wvf/++3HwwQfv7jYDAABAr1dVcA8YMKC8zdecOXNixYoVsWTJkmhra4upU6d27u3eunVr+fHZZ58dDz/8cDz++OPxwQcflIeYF3u3J02alPM7AQAAgF6kqkPKC8WFz4rgPv/888vbfM2cOTNOPvnk8nOtra1xyy23xOTJk8urlH/xxRfllcn//ve/l3vH58+fX9UF0wAAAGBv1VCp9O4zpIuLSbloGrV88ZGWlkGxdq2LplG7zDn1wJxTD8w59aCxsXuvxF/VIeUAAADArhHcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAIDgBgAAgL2DPdwAAACQQHADAABAbwjubdu2xaxZs2LcuHHR2toabW1tX7vu22+/Heecc04ce+yxcfrpp8crr7yyu9sLAAAAtRncc+fOjZUrV8b8+fNj9uzZMW/evFi8ePFO623atCkuvPDCGDFiRDz55JMxfvz4uOSSS2LdunXdte0AAABQG8G9efPmWLhwYVxzzTUxatSoMqKnT58eCxYs2Gndxx57LPbdd9+YM2dOHHrooXHppZeW74tYBwAAgFrXp5qVV61aFTt27IgxY8Z0Lhs7dmzcc8890d7eHo2N/9/vr732Wpx00kmxzz77dC575JFHumu7AQAAoHaCe82aNTFkyJDo27dv57KWlpbyvO6NGzdGU1NT5/KPPvqoPHf72muvjeeffz4OOuiguOqqq8pAr0ZDw7/eoBZ1zLYZp5aZc+qBOacemHPqQUNDDwb3li1busR2oePx9u3bdzr8/L777oupU6fG/fffH3/5y19i2rRp8fTTT8e3vvWtXf6aTU2DqtlE2Cs1N5tzap85px6Yc+qBOYek4O7Xr99OYd3xuH///l2WF4eSjxw5sjx3u3D00UfHiy++GE888UT88pe/3OWvuX79pmhvr2YrYe96Ba34S2vduk1RqfT01kAOc049MOfUA3NOPWhs7N6dvlUF97Bhw2LDhg3ledx9+vTpPMy8iO3Bgwd3WffAAw+M73znO12WHXbYYfHpp59WtYFFhAgRap05px6Yc+qBOacemHNqWaXSg1cpL/ZYF6G9bNmyzmVLly6N0aNHd7lgWuH4448v78P9Ze+99155LjcAAADUuqqCe8CAATFx4sTyVl8rVqyIJUuWRFtbW3medsfe7q1bt5YfT5kypQzuu+++Oz744IO48847ywupnXHGGTm/EwAAANhbg7tw9dVXl/fgPv/88+P666+PmTNnxsknn1x+rrW1NRYtWlR+XOzJ/sMf/hB//etf47TTTivfFxdRKw5LBwAAgFrXUKn07jOki4tJuWgatXzxkZaWQbF2rYumUbvMOfXAnFMPzDn1oLGxe6/EX/UebgAAAOC/E9wAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAgOAGAACAvYM93AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEBvCO5t27bFrFmzYty4cdHa2hptbW3/9Tkff/xxjBkzJl599dX/dTsBAABgr9Kn2ifMnTs3Vq5cGfPnz49PPvkkrrrqqvj2t78dEyZM+NrnzJkzJzZv3ry72woAAAC1GdxFNC9cuDDuv//+GDVqVPm2evXqWLBgwdcG95///Of44osvumt7AQAAoPYOKV+1alXs2LGjPDy8w9ixY2P58uXR3t6+0/obNmyI22+/PW644Ybu2VoAAACoxT3ca9asiSFDhkTfvn07l7W0tJTndW/cuDGampq6rH/rrbfGpEmT4vDDD/+fN7Ch4V9vUIs6ZtuMU8vMOfXAnFMPzDn1oKGhB4N7y5YtXWK70PF4+/btXZa/9NJLsXTp0njqqad2awObmgbt1vNhb9DcbM6pfeacemDOqQfmHJKCu1+/fjuFdcfj/v37dy7bunVrXHfddTF79uwuy/8X69dviv9wtDrUzCtoxV9a69Ztikqlp7cGcphz6oE5px6Yc+pBY2P37vStKriHDRtWnpddnMfdp0+fzsPMi6gePHhw53orVqyIjz76KC699NIuz7/oooti4sSJVZ3TXUSIEKHWmXPqgTmnHphz6oE5p5ZVunknWFXBPXLkyDK0ly1bVt6Hu1AcNj569OhoLF4K+Ldjjz02nnnmmS7PPfnkk+Omm26K73//+9217QAAANBrVRXcAwYMKPdQF/fV/u1vfxv/+Mc/oq2tLW655ZbOvd2DBg0q93gfeuih/3EPeXNzc/dtPQAAANTCbcEKV199dXn/7fPPPz+uv/76mDlzZrn3utDa2hqLFi3K2E4AAADYqzRUKr37DOniYlIumkYtX3ykpWVQrF3romnULnNOPTDn1ANzTj1obOzeK/FXvYcbAAAA+O8ENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAA9Ibg3rZtW8yaNSvGjRsXra2t0dbW9rXrvvDCC3HGGWfEmDFj4vTTT4/nnntud7cXAAAAajO4586dGytXroz58+fH7NmzY968ebF48eKd1lu1alVccsklceaZZ8bjjz8eU6ZMicsuu6xcDgAAALWuTzUrb968ORYuXBj3339/jBo1qnxbvXp1LFiwICZMmNBl3aeeeiq+973vxdSpU8vHhx56aDz//PPx9NNPx1FHHdW9vwsAAADYm4O72Du9Y8eO8hDxDmPHjo177rkn2tvbo7Hx/3eYT5o0Kf75z3/u9Gts2rRpd7cZAAAAaiu416xZE0OGDIm+fft2LmtpaSnP6964cWM0NTV1Lh8+fHiX5xZ7wl9++eXy0PJqNDT86w1qUcdsm3FqmTmnHphz6oE5px40NPRgcG/ZsqVLbBc6Hm/fvv1rn7d+/fqYOXNmnHDCCXHSSSdVtYFNTYOqWh/2Rs3N5pzaZ86pB+acemDOISm4+/Xrt1NYdzzu37//f3zO2rVr4xe/+EVUKpW46667uhx2vivWr98U7e1VPQX2qlfQir+01q3bFJVKT28N5DDn1ANzTj0w59SDxsbu3elbVXAPGzYsNmzYUJ7H3adPn87DzIvYHjx48E7rf/bZZ50XTXvggQe6HHK+q4oIESLUOnNOPTDn1ANzTj0w59SySjfvBKtqd/PIkSPL0F62bFnnsqVLl8bo0aN32nNdXNF8+vTp5fIHH3ywjHUAAACoF1UF94ABA2LixIkxZ86cWLFiRSxZsiTa2to692IXe7u3bt1afnzvvffGhx9+GLfddlvn54o3VykHAACgHjRUipOrq7xwWhHczzzzTAwcODCmTZsWF1xwQfm5I488Mm655ZaYPHlyeV/u999/f6fnF7cLu/XWW3f56xXntjqHm1o+F6qlZVCsXescbmqXOacemHPqgTmnHjQ2du+FAasO7j1NcFPL/MVFPTDn1ANzTj0w59SDxm4O7uouGQ4AAADsEsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAAghsAAAD2DvZwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAAkENwAAACQQHADAABAAsENAAAACQQ3AAAAJBDcAAAAkEBwAwAAQALBDQAAAAkENwAAACQQ3AAAAJBAcAMAAEACwQ0AAAAJBDcAAAD0huDetm1bzJo1K8aNGxetra3R1tb2teu+9dZbcdZZZ8Vxxx0XZ555ZqxcuXJ3txcAAABqM7jnzp1bhvP8+fNj9uzZMW/evFi8ePFO623evDlmzJhRhvmjjz4aY8aMiYsvvrhcDgAAALWuquAuYnnhwoVxzTXXxKhRo2L8+PExffr0WLBgwU7rLlq0KPr16xdXXnllDB8+vHzOfvvt9x/jHAAAAOo6uFetWhU7duwo91Z3GDt2bCxfvjza29u7rFssKz7X0NBQPi7en3DCCbFs2bLu2nYAAADotfpUs/KaNWtiyJAh0bdv385lLS0t5XndGzdujKampi7rjhgxosvzm5ubY/Xq1VVtYNHrjS7tRo369+tR5YxXKj29NZDDnFMPzDn1wJxTT3PeI8G9ZcuWLrFd6Hi8ffv2XVr3q+v9N01Ng6paH/ZG5px6YM6pB+acemDOYddVte+4OCf7q8Hc8bh///67tO5X1wMAAICo9+AeNmxYbNiwoTyP+8uHjhcRPXjw4J3WXbt2bZdlxeOhQ4fu7jYDAABAbQX3yJEjo0+fPl0ufLZ06dIYPXp0NH7lROvi3ttvvPFGVP59Ymrx/vXXXy+XAwAAQK2rKrgHDBgQEydOjDlz5sSKFStiyZIl0dbWFlOnTu3c271169by4wkTJsTnn38eN998c7zzzjvl++K87lNOOSXndwIAAAC9SEOlYxf0LiqiuQjuZ555JgYOHBjTpk2LCy64oPzckUceGbfccktMnjy5fFxE+ezZs+Pdd98tP3f99dfH0UcfnfM7AQAAgL05uAEAAID/zh2uAQAAIIHgBgAAgASCGwAAAGotuLdt2xazZs2KcePGRWtra3nF86/z1ltvxVlnnVXeVuzMM8+MlStX7tFthT0x5y+88EKcccYZMWbMmDj99NPjueee8wdPzc15h48//ric9VdffXWPbCPsyTl/++2345xzzoljjz22/Hn+yiuv+B9Azc35s88+W96BqPhZXsz7m2++uUe3FXbX9u3b47TTTvvGf4vsbof2aHDPnTu33OD58+eXVzOfN29eLF68eKf1Nm/eHDNmzCi/8R999NHym/riiy8ul0Nvt6tzvmrVqrjkkkvKb+THH388pkyZEpdddlm5HGplzr+suOOFn+PU4pxv2rQpLrzwwhgxYkQ8+eSTMX78+PLn+7p163pkuyFjzlevXh1XXHFF+W/yJ554IkaOHFl+XNzRCPaWF5cuv/zycpa/Trd0aKWHfPHFF5XRo0dXXnnllc5lv//97ys/+9nPdlp34cKFlR/+8IeV9vb28nHxfvz48ZVHHnlkj24zZM757bffXpk2bVqXZRdeeGHld7/7nT94ambOOzzxxBOVKVOmVI444oguz4NamPP58+dXfvSjH1V27NjRuWzy5MmVF154YY9tL2TP+R//+MfKpEmTOh9v2rSp/Jm+YsUKf/j0eqtXr6785Cc/qZx++unf+G+R7ujQHtvDXey127FjR/kqQYexY8fG8uXLo729vcu6xbLicw0NDeXj4v0JJ5wQy5Yt2+PbDVlzPmnSpPj1r3/9H/eUQK3MeWHDhg1x++23xw033LCHtxT2zJy/9tprcdJJJ8U+++zTueyRRx6JH/zgB/4XUDNzfsABB8Q777wTS5cuLT9X7P0bOHBgHHLIIT2w5VCd4uf0iSeeGA899NA3rtcdHdonesiaNWtiyJAh0bdv385lLS0t5a79jRs3RlNTU5d1i8Oyvqy5ufkbd/9Db1DNnA8fPrzLc4v5fvnll8tDy6FW5rxw6623li8wHX744T2wtZA/5x999FF57va1114bzz//fBx00EFx1VVXlf9og1qZ81NPPbWc73PPPbd8camxsTHuvffe2H///Xto62HXFXO7K7qjQ3tsD3dxfseXv5kLHY+Lk9d3Zd2vrge9TTVz/mXr16+PmTNnlq+gFXtJoFbm/KWXXir3hvzqV7/ao9sIe3LOi3P77rvvvjjwwAPj/vvvj+9+97sxbdq0+PTTT/2PoGbmvDhaqYiR6667Lh5++OHyoq9XX321axVQU7Z0Q4f2WHD369dvpw3teNy/f/9dWver60FvU82cd1i7dm2cf/75xfUV4q677ipfMYZamPOtW7eW/zArLsLj5ze1/PO82NtXXEDq0ksvjaOPPjp+85vfxGGHHVZeWApqZc7vuOOOOOKII+K8886LY445Jm688cYYMGBAefoE1Ip+3dChPfYv+WHDhpWvjBXniXQoXiUrNn7w4ME7rVtEyJcVj4cOHbrHthey57zw2WeflX9xFd/IDzzwwE6H4sLePOcrVqwoD7UtIqQ4P7DjHMGLLrqoDHGolZ/nxZ7t73znO12WFcFtDze1NOfFLcCOOuqozsfFDoLi8SeffLJHtxkydUeH9lhwF6/89unTp8sJ58VhhqNHj95pj15xz7M33nij3ONXKN6//vrr5XLozaqZ8+IQxOnTp5fLH3zwwfIbHGppzotzWp955pnytncdb4WbbrqpvAUe1MrP8+OPP768D/eXvffee+W53FArc14Ex7vvvttl2fvvvx8HH3zwHtteyNYdHdpjwV0ccjJx4sTyPqzFXo8lS5ZEW1tbTJ06tfPVtOLww8KECRPi888/j5tvvrm8GmLxvjie/pRTTumpzYdun/PiQiMffvhh3HbbbZ2fK95cpZxamfNiD8mhhx7a5a1QvLhUXIAEauXneXGxyyK477777vjggw/izjvvLI/uKM5xhVqZ87PPPrs8d7t48bSY8+IQ82LvdnFRTNibrenuDq30oM2bN1euvPLKyvHHH19pbW0t7+fXobgf2pfvb7Z8+fLKxIkTy3sD/vSnP628+eabPbTVkDPnP/7xj8vHX3276qqr/JHT61Xz8/zL3IebWp3zv/3tb+U9io855pjKGWecUXnttdd6aKshb84ffvjhyoQJE8p1zznnnMrKlSv9cbPXOeIr9+Hu7g5tKP6T9/oAAAAA1CeXPwYAAIAEghsAAAASCG4AAABIILgBAAAggeAGAACABIIbAAAAEghuAAAASCC4AQAAIIHgBgAAgASCGwAAABIIbgAAAEgguAEAACC63/8BO1dfc5wg/PUAAAAASUVORK5CYII="
     },
     "metadata": {},
     "output_type": "display_data",
     "jetTransient": {
      "display_id": null
     }
    }
   ],
   "execution_count": 5
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Summary Statistics",
   "id": "2a0add510675e583"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Create summary report\n",
    "with open('./results/analysis_summary.txt', 'w') as f:\n",
    "    f.write(\"SARS-CoV-2 Differential Expression Analysis Summary\\n\")\n",
    "    f.write(\"=\"*60 + \"\\n\\n\")\n",
    "    f.write(f\"Dataset: GSE147507\\n\")\n",
    "    f.write(f\"Total genes analyzed: {len(de_df)}\\n\\n\")\n",
    "    f.write(\"Differentially Expressed Genes (|Log2FC| > 1, adj.p < 0.05):\\n\")\n",
    "    f.write(f\"  - Upregulated: {sum(de_df['Regulation'] == 'Up')}\\n\")\n",
    "    f.write(f\"  - Downregulated: {sum(de_df['Regulation'] == 'Down')}\\n\")\n",
    "    f.write(f\"  - Not significant: {sum(de_df['Regulation'] == 'Not Significant')}\\n\\n\")\n",
    "\n",
    "    f.write(\"Top 10 Upregulated Genes:\\n\")\n",
    "    top_up = de_df[de_df['Regulation'] == 'Up'].nlargest(10, 'Log2FC')\n",
    "    for _, row in top_up.iterrows():\n",
    "        f.write(f\"  {row['Gene']}: Log2FC={row['Log2FC']:.2f}, adj.p={row['Adjusted_P']:.2e}\\n\")\n",
    "\n",
    "    f.write(\"\\nTop 10 Downregulated Genes:\\n\")\n",
    "    top_down = de_df[de_df['Regulation'] == 'Down'].nsmallest(10, 'Log2FC')\n",
    "    for _, row in top_down.iterrows():\n",
    "        f.write(f\"  {row['Gene']}: Log2FC={row['Log2FC']:.2f}, adj.p={row['Adjusted_P']:.2e}\\n\")\n",
    "\n",
    "print(\"Summary report saved!\")"
   ],
   "id": "f2cc32abffac5fec",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "print(f\"Expression data shape: {expr_log2.shape}\")\n",
    "print(f\"Sample columns: {expr_log2.columns.tolist()}\")\n",
    "print(f\"Any data? {expr_log2.values.sum()}\")"
   ],
   "id": "f85b8eed009d18f7",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "print(f\"Infected samples ({len(infected_cols)}): {infected_cols}\")\n",
    "print(f\"Mock samples ({len(mock_cols)}): {mock_cols}\")\n",
    "print(\"\\nRegulation counts:\")\n",
    "print(de_df['Regulation'].value_counts())\n",
    "print(f\"\\nHow many genes with adj.p < 0.05? {sum(de_df['Adjusted_P'] < 0.05)}\")\n",
    "print(f\"How many genes with |Log2FC| > 1? {sum(abs(de_df['Log2FC']) > 1)}\")"
   ],
   "id": "9655ad6a67ccf8b",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
